contracts/reverseRegistrar/L2ReverseRegistrar.sol

```diff
@@ -0,0 +1,329 @@
+pragma solidity >=0.8.4;
+
+import "../registry/ENS.sol";
+import "./IL2ReverseRegistrar.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/cryptography/SignatureChecker.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "../root/Controllable.sol";
+import "./profiles/L2NameResolver.sol";
+import "./profiles/L2TextResolver.sol";
+import "./profiles/L2ReverseResolverBase.sol";
+
+error InvalidSignature();
+
+contract L2ReverseRegistrar is
+    Ownable,
+    IL2ReverseRegistrar,
+    L2ReverseResolverBase,
+    L2NameResolver,
+    L2TextResolver
+{
+    using ECDSA for bytes32;
+    mapping(bytes32 => uint256) public lastUpdated;
+
+    event ReverseClaimed(address indexed addr, bytes32 indexed node);
+    event DefaultResolverChanged(L2NameResolver indexed resolver);
+
+    /**
+     * @dev Constructor
+     */
+    constructor(bytes32 L2ReverseNode) L2ReverseResolverBase(L2ReverseNode) {}
+
+    modifier authorised(address addr) override(L2ReverseResolverBase) {
+        isAuthorised(addr);
+        _;
+    }
+
+    function isAuthorised(address addr) internal view override returns (bool) {
+        require(
+            addr == msg.sender || ownsContract(addr, msg.sender),
+            "ReverseRegistrar: Caller is not a controller or authorised by address or the address itself"
+        );
+    }
+
+    /**
+     * @dev Sets the name for an addr using a signature that can be verified with ERC1271.
+     * @param addr The reverse record to set
+     * @param name The name of the reverse record
+     * @param inceptionDate Date from when this signature is valid from
+     * @param signature The resolver of the reverse node
+     * @return The ENS node hash of the reverse record.
+     */
+    function setNameForAddrWithSignature(
+        address addr,
+        string memory name,
+        uint256 inceptionDate,
+        bytes memory signature
+    ) public override returns (bytes32) {
+        bytes32 node = _getNamehash(addr);
+
+        bytes32 hash = keccak256(
+            abi.encodePacked(
+                IL2ReverseRegistrar.setNameForAddrWithSignature.selector,
+                addr,
+                name,
+                inceptionDate
+            )
+        );
+
+        bytes32 message = hash.toEthSignedMessageHash();
+
+        if (
+            !SignatureChecker.isValidSignatureNow(addr, message, signature) ||
+            inceptionDate <= lastUpdated[node] ||
```

Comment:
Is this check necessary assuming `lastUpdated[node]` is always > `block.timestamp` ?

Reply:
Wouldn't `lastUpdated[node]` always be < `block.timestamp`

Reply:
I think this has changed since and I've added comments, so maybe worth looking at again